### PR TITLE
http: added a new dual header mutation filter that could be used as downstream/upstream filter

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: http: added a new dual header mutation filter that could be used as downstream/upstream filter
Additional Description:

Check #24100 for more detailed context.

Risk Level: low. new L7 extension.
Testing: unit.
Docs Changes: added.
Release Notes: added.
Platform Specific Features: n/a.

### CODE REVIEW

**Comment by wbpcode** [comment]
cc @alyssawilk here is the pr for #24100. This PR is consist of three commits. 

The first commit added a new `HeaderEvaluator` implementation to the `common/http`. The second PR refactor previous `early_header_mutation` to use the new implementation in the `common/http`. The final commit create a new L7 filter.

File: CODEOWNERS
Comment (htuch) [inline CODEOWNERS:298]: Should this be starting life without an additional maintainer to provide coverage?

File: api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto:27]: What is this for?

File: api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto:21]: Does the decoder/encoder terminology still make sense when this is used as an upstream filter?

File: docs/root/configuration/http/http_filters/header_mutation_filter.rst
Comment (htuch) [inline docs/root/configuration/http/http_filters/header_mutation_filter.rst:6]: How should an operator think about this in relation to the early HTTP mutation filter that already exists? Is this a strict superset an basically obviates it?

File: CODEOWNERS
Comment (wbpcode) [reply CODEOWNERS:298]: The code/logic of this filter is very simple. I am not sure if it's necessary to have an additional maintainer to help? 🤔

File: api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto
Comment (wbpcode) [reply api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto:27]: Maybe my naming is too ambiguous.

The `HeaderMutation` here is a per route filter proto API. And `HeaderMutationConfig` is a empty filter proto API.

May be using `HeaderMutationPerRoute` and `HeaderMutation` is better?

File: docs/root/configuration/http/http_filters/header_mutation_filter.rst
Comment (wbpcode) [reply docs/root/configuration/http/http_filters/header_mutation_filter.rst:6]: This would be a more flexible alternative to the tranditional `headers_to_add`/`headers_to_remove` in most cases.

However, this still be a normal HTTP filter. In some corner cases, this filter cann't be used as alternative of `headers_to_add`/`headers_to_remove`. For example, if `sendDirectLocalReply` is called, then the encoder filters will be skipped.

File: api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto
Comment (wbpcode) [reply api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto:21]: > Does the decoder/encoder terminology still make sense when this is used as an upstream filter?

May be request/response is better?

The `decodeHeaders` will be used to handle the **request headers** in both downstream http filter and upstream http filter for now. Although I personlly think that `decodeHeaders` should be used to handle the **response headers** in upstream http filter.

File: CODEOWNERS
Comment (htuch) [reply CODEOWNERS:298]: Sure, but bus factor of one isn't great for ownership and https://github.com/envoyproxy/envoy/blob/main/EXTENSION_POLICY.md suggests two.

File: api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto
Comment (htuch) [reply api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto:27]: SG with route in the name and explaining this in the comments.

File: api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto
Comment (htuch) [reply api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto:21]: I think request/response makes more sense TBH.

File: docs/root/configuration/http/http_filters/header_mutation_filter.rst
Comment (htuch) [reply docs/root/configuration/http/http_filters/header_mutation_filter.rst:6]: OK, maybe inline this explanation.

File: CODEOWNERS
Comment (htuch) [reply CODEOWNERS:298]: Also we could have this other maintainer review this initial implementation to spin them up.

Comment (wbpcode) [reply CODEOWNERS:298]: Get it. I wonder do you have free bandwidth for this? 😸  If you have free time, then we needn't to bother third maintainer. Thanks.

Comment (htuch) [reply CODEOWNERS:298]: Yeah, I actually have a need to deep dive in some related parts of Envoy, so I can do this if there isn't a more interested party :)

File: docs/root/configuration/http/http_filters/header_mutation_filter.rst
Comment (htuch) [inline docs/root/configuration/http/http_filters/header_mutation_filter.rst:18]: One of the interesting properties of this design, which is probably worth spelling out, is that if a later filter clears the route cache, the mutated headers will influence the new route pick. That was never the case for `request_headers_to_add` etc. I think this makes the new filter quite a bit more useful in some configurations.

**Review by htuch — CHANGES_REQUESTED** [review]
Looks good at API / design level! Some minor docs nits and a couple of Qs.

File: docs/root/configuration/http/http_filters/header_mutation_filter.rst
Comment (wbpcode) [reply docs/root/configuration/http/http_filters/header_mutation_filter.rst:20]: it's a corner case. If one encoder filter stops the filter chain and calls the send local reply, then the local response will bypass the encoder filter chain.

But it's too corner. Maybe we can remove this note directly?

File: docs/root/configuration/http/http_filters/header_mutation_filter.rst
Comment (wbpcode) [reply docs/root/configuration/http/http_filters/header_mutation_filter.rst:18]: This is oneof the target scenes. (And I think may be we can add a `clear_route_cache` option in the filter config in the future? 🤔

File: api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto
Comment (wbpcode) [reply api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto:19]: I think the `route` is the most specific level for filter configuration. And for the upstream filters, it still should be possible to apply diffferent configurations for different routes that refer to the same cluster.

File: docs/root/configuration/http/http_filters/header_mutation_filter.rst
Comment (htuch) [reply docs/root/configuration/http/http_filters/header_mutation_filter.rst]: I'd just say "Please note that as an encoder filter, this filter follows the standard rules of when it will execute in situations such as local replies - response headers will not be unconditionally added in cases where the filter would be bypassed."

File: docs/root/configuration/http/http_filters/header_mutation_filter.rst
Comment (htuch) [reply docs/root/configuration/http/http_filters/header_mutation_filter.rst:16]: "Header mutations may be used to influence later route picks if the route cache is cleared by a filter executing after the header mutation filter"

Yeah, I think a clear_route_cache option could make sense when there is a use case. I'd split it to a different PR.

File: api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto
Comment (htuch) [reply api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto:19]: Sure, it is most specific. But, what if I just want to set header mutation for the whole cluster?

Comment (wbpcode) [reply api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto:19]: Then, non-empty filter configuraion (`HeaderMutation` in the proto) would be a choice.

File: docs/root/configuration/http/http_filters/header_mutation_filter.rst
Comment (htuch) [inline docs/root/configuration/http/http_filters/header_mutation_filter.rst:13]: Is it fair to say this alternative use would sometimes also be done via composite filter?

File: source/extensions/filters/http/header_mutation/header_mutation.cc
Comment (htuch) [inline source/extensions/filters/http/header_mutation/header_mutation.cc:36]: Do you want to do both per-route and filter-level config, or have route-specific override the filter config?

File: source/extensions/filters/http/header_mutation/header_mutation.cc
Comment (htuch) [inline source/extensions/filters/http/header_mutation/header_mutation.cc:54]: Why is this empty? Don't we know the request headers from earlier?

File: test/common/http/header_mutation_test.cc
Comment (htuch) [inline test/common/http/header_mutation_test.cc:1]: Do you have any integration tests, in particular ones that cover the upstream filter use case?

Comment (wbpcode) [reply test/common/http/header_mutation_test.cc:1]: I will add some integration tests for this filter.

File: source/extensions/filters/http/header_mutation/header_mutation.cc
Comment (wbpcode) [reply source/extensions/filters/http/header_mutation/header_mutation.cc:54]: In fact, logically to say, the request headers will never be nullptr.

I will update these code to use a ASSERT here to avoid the nullptr check.

File: docs/root/configuration/http/http_filters/header_mutation_filter.rst
Comment (wbpcode) [reply docs/root/configuration/http/http_filters/header_mutation_filter.rst:13]: What do you mean by the `composite filter` here? The `envoy.filters.http.composite` filter or multiple header mutation filters?

File: source/extensions/filters/http/header_mutation/header_mutation.cc
Comment (htuch) [reply source/extensions/filters/http/header_mutation/header_mutation.cc:36]: Maybe call this out in docs / comments somewhere.

File: docs/root/configuration/http/http_filters/header_mutation_filter.rst
Comment (htuch) [reply docs/root/configuration/http/http_filters/header_mutation_filter.rst:13]: envoy.filters.http.composite

File: test/extensions/filters/http/header_mutation/header_mutation_integration_test.cc
Comment (htuch) [inline test/extensions/filters/http/header_mutation/header_mutation_integration_test.cc:47]: It's a bit jarring to have both YAML and proto generated methods used depending on downstream/upstream. If config helper is missing support for upstream filters, maybe now is the time to add it.

File: test/extensions/filters/http/header_mutation/header_mutation_integration_test.cc
Comment (htuch) [inline test/extensions/filters/http/header_mutation/header_mutation_integration_test.cc:26]: Is there value in also covering request mutations?

File: source/extensions/filters/http/header_mutation/header_mutation.cc
Comment (wbpcode) [reply source/extensions/filters/http/header_mutation/header_mutation.cc:36]: I think I has called this out in the proto API.

File: docs/root/configuration/http/http_filters/header_mutation_filter.rst
Comment (wbpcode) [reply docs/root/configuration/http/http_filters/header_mutation_filter.rst:13]: I think the answer is yes. The compositer filter could provides very flexible matching. Although for most users, route level configuration is flexible enough.

File: test/extensions/filters/http/header_mutation/header_mutation_integration_test.cc
Comment (wbpcode) [reply test/extensions/filters/http/header_mutation/header_mutation_integration_test.cc:47]: ok, I will add new helper function for the upstream http filters.

File: test/extensions/filters/http/header_mutation/header_mutation_integration_test.cc
Comment (wbpcode) [reply test/extensions/filters/http/header_mutation/header_mutation_integration_test.cc:26]: get it. I will add some request mutations here.

File: test/extensions/filters/http/header_mutation/header_mutation_integration_test.cc
Comment (wbpcode) [reply test/extensions/filters/http/header_mutation/header_mutation_integration_test.cc]: Seems that `prependFilter` is enough for this requirement.

File: api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto:33]:
```suggestion
// always be applied first and then the per-route mutation rules, if both are specified.
```

File: test/extensions/http/early_header_mutation/header_mutation/header_mutation_test.cc
Comment (htuch) [inline test/extensions/http/early_header_mutation/header_mutation/header_mutation_test.cc:62]: Why get rid of all these tests? I realize we want to have this newer filter subsume many of the cases of the early header mutation filter, but not sure why the selective removal of tests in the current PR.

File: source/extensions/filters/http/header_mutation/header_mutation.cc
Comment (htuch) [reply source/extensions/filters/http/header_mutation/header_mutation.cc:36]: Yeah, I made a small suggestion to make it clearer but you are right.

**Comment by htuch** [comment]
LGTM modulo comment, thanks, this is really nice.

File: test/extensions/http/early_header_mutation/header_mutation/header_mutation_test.cc
Comment (wbpcode) [reply test/extensions/http/early_header_mutation/header_mutation/header_mutation_test.cc:62]: In fact, they ware not removed but move to the https://github.com/envoyproxy/envoy/pull/25658/files#diff-e87c3085497f9392ed99f92ff4d3189cd16810be1574517625f72a08d75b6d7c.

We extracted some common code to the `common/http/header_mutation`, so all these old tests are moved to the common part.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

File: source/extensions/filters/http/header_mutation/header_mutation.cc
Comment (botengyao) [inline source/extensions/filters/http/header_mutation/header_mutation.cc:56]: Can we relax this assert here? Seems it is breaking [filter_fuzz_test](https://github.com/envoyproxy/envoy/blob/30cfa5c4619bae49125fff35aacf2f62e81b4410/test/extensions/filters/http/common/fuzz/BUILD#L66) with this corpus:
```
config {
  name: "envoy.filters.http.header_mutation"
  typed_config {
    type_url: "type.googleapis.com/envoy.extensions.filters.http.header_mutation.v3.HeaderMutation"
  }
}
```
Or we may need to change the fuzz test infra? Can you take a look? Thanks!

Comment (wbpcode) [reply source/extensions/filters/http/header_mutation/header_mutation.cc:56]: I think we need a fix here. I just get some new knowledge about the filter chain. It's possible to get a null request headers in the encoding phase.

Comment (botengyao) [reply source/extensions/filters/http/header_mutation/header_mutation.cc:56]: Thanks!

**Pull request merged**
